File: app/src/main/java/com/example/cookbook2/FavouriteRecipe/FavouriteRecipeActivity.java
package com.example.cookbook2.FavouriteRecipe;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import com.example.cookbook2.FavouriteRecipeModel.FavouriteRecipe;
import com.example.cookbook2.FavouriteRecipeModel.FavouriteRecipeViewModel;
import com.example.cookbook2.R;
import com.example.cookbook2.Recipe.JsonData2Recipe;
import com.example.cookbook2.Recipe.RecipeModel;
import com.example.cookbook2.RecipeAdapter;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class FavouriteRecipeActivity extends AppCompatActivity {

    private final static String TAG = FavouriteRecipeActivity.class.getName();

    private FavouriteRecipeViewModel fRViewModel;
    private RecyclerView rv;
    private RecipeAdapter adapter;


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.favourite_recipe_menu,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if(item.getItemId()==R.id.delete_all_recipe){
            fRViewModel.deleteAll();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_favourite_recipe);
        setTitle("Favourite Recipes");

        rv = findViewById(R.id.favourite_recipe_rv);
        adapter = new RecipeAdapter(this, new ArrayList<RecipeModel>());
        rv.setAdapter(adapter);
        rv.setLayoutManager(new LinearLayoutManager(FavouriteRecipeActivity.this, RecyclerView.VERTICAL, false));
        fRViewModel = new ViewModelProvider(this).get(FavouriteRecipeViewModel.class);
        Disposable d =  fRViewModel.getAll().subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<List<FavouriteRecipe>>() {
                    @Override
                    public void accept(List<FavouriteRecipe> favouriteRecipes) throws Exception {
                        List<RecipeModel> recipeList = new ArrayList<>();
                        for(int i=0; i<favouriteRecipes.size();i++){
                            FavouriteRecipe recipe = favouriteRecipes.get(i);
                            String url = JsonData2Recipe.subImageUrl1+recipe.getId()+JsonData2Recipe.subImageUrl2;
                            Log.d(TAG, url);
                            RecipeModel r = new RecipeModel(recipe.getId(),recipe.getTitle(), url, 0, true, null);
                            recipeList.add(r);
                        }
                        adapter.modifyData(recipeList);
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        throwable.printStackTrace();
                    }
                });
            /*
                observe(this, new Observer<List<FavouriteRecipe>>() {
            @Override
            public void onChanged(List<FavouriteRecipe> favouriteRecipes) {


            }
        });*/

    }



}

File: app/src/main/java/com/jasonstudio/cookbook2/view/activity/FavouriteRecipeActivity.kt
package com.jasonstudio.cookbook2.view.activity

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.jasonstudio.cookbook2.FavouriteRecipeModel.FavouriteRecipeViewModel
import com.jasonstudio.cookbook2.R
import com.jasonstudio.cookbook2.Recipe.JsonData2Recipe
import com.jasonstudio.cookbook2.Recipe.RecipeModel
import com.jasonstudio.cookbook2.RecipeAdapter
import com.jasonstudio.cookbook2.databinding.ActivityFavouriteRecipeBinding
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

class FavouriteRecipeActivity : BaseActivity<ActivityFavouriteRecipeBinding>(ActivityFavouriteRecipeBinding::inflate) {
    private lateinit var fRViewModel: FavouriteRecipeViewModel
    private lateinit var adapter: RecipeAdapter
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.favourite_recipe_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.delete_all_recipe) {
            fRViewModel.deleteAll()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        title = "Favourite Recipes"
        adapter = RecipeAdapter(this, ArrayList())
        binding.favouriteRecipeRv.apply {
            adapter = this@FavouriteRecipeActivity.adapter
            layoutManager = LinearLayoutManager(
                this@FavouriteRecipeActivity,
                RecyclerView.VERTICAL,
                false
            )
        }
        fRViewModel = ViewModelProvider(this)[FavouriteRecipeViewModel::class.java]
        val d = fRViewModel.all.subscribeOn(Schedulers.io())
            ?.observeOn(AndroidSchedulers.mainThread())
            ?.subscribe({ favouriteRecipes ->
                val recipeList: MutableList<RecipeModel> = ArrayList()
                for (i in favouriteRecipes.indices) {
                    val recipe = favouriteRecipes[i]
                    val url =
                        JsonData2Recipe.subImageUrl1 + recipe.id + JsonData2Recipe.subImageUrl2
                    Log.d(TAG, url)
                    val r = RecipeModel(recipe.id, recipe.title, url, 0, true, null)
                    recipeList.add(r)
                }
                adapter.modifyData(recipeList)
            }, { throwable -> throwable.printStackTrace() })
    }

    companion object {
        private val TAG = FavouriteRecipeActivity::class.java.name
    }
}